src/sinks/util/service/concurrency.rs

```diff
@@ -0,0 +1,111 @@
+use serde::de::{self, Unexpected, Visitor};
+use serde::{Deserialize, Deserializer, Serialize};
+use std::fmt;
+
+pub trait ConcurrencyOption {
+    fn parse_concurrency(&self, default: &Self) -> Option<usize>;
+    fn is_none(&self) -> bool;
+    fn is_some(&self) -> bool {
+        !self.is_none()
+    }
+}
+
+impl ConcurrencyOption for Option<usize> {
+    fn parse_concurrency(&self, default: &Self) -> Option<usize> {
+        let limit = match self {
+            None => *default,
+            Some(x) => Some(*x),
+        };
+        limit.or(Some(1024))
```

Comment:
I think this is equivalent to:
```suggestion
        self.or(*default).map(|x| *x).or(Some(1024))
```